examples/ft-refund/solidity-contract/src/FtRefund.sol

```diff
@@ -0,0 +1,170 @@
+            address(this.wNEAR()),
+            abi.encodeWithSelector(
+                0x095ea7b3, // approve method selector
+                address(this),
+                amount
+            ),
+            0,
+            APPROVE_NEAR_GAS
+        );
+        approveCall.transact();
+    }
+
+    function ftTransferCall(
+        IEvmErc20 token,
+        string memory tokenId,
+        uint128 amount
+    ) public {
+        token.transferFrom(msg.sender, address(this), amount);
+        token.withdrawToNear(
+            abi.encodePacked(AuroraSdk.nearRepresentative(address(this))),
+            uint(amount)
+        );
+
+        bytes memory data = abi.encodePacked(
+            "{",
+            '"receiver_id": "',
+            nearAccountId,
+            '",',
+            '"amount": "',
+            Strings.toString(amount),
+            '",',
+            '"msg": "refund"',
+            "}"
+        );
+        PromiseCreateArgs memory callFtTransfer = near.call(
+            tokenId,
+            "ft_transfer_call",
+            data,
+            1,
+            FT_TRANSFER_CALL_NEAR_GAS
+        );
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(
+                this.ftTransferCallCallback.selector,
+                msg.sender,
+                tokenId,
+                amount
+            ),
+            0,
+            CALLBACK_NEAR_GAS
+        );
+
+        callFtTransfer.then(callback).transact();
+    }
+
+    function ftTransferCallCallback(
+        address sender,
+        string memory tokenIdOnNear,
+        uint128 amount
+    ) public onlyRole(CALLBACK_ROLE) {
+        PromiseResult memory promiseResult = AuroraSdk.promiseResult(0);
+        uint128 refundAmount = 0;
+
+        if (promiseResult.status != PromiseResultStatus.Successful) {
+            // if Promise failed we need to do whole refund
+            refundAmount = amount;
+        } else {
+            // else `ft_on_transfer` will return used amount of FT,
+            // which we need to extract from original amount
+            uint128 usedAmount = _stringToUint(string(promiseResult.output));
+            refundAmount = amount - usedAmount;
+        }
+
+        if (refundAmount > 0) {
+            bytes memory data = abi.encodePacked(
+                "{",
+                '"receiver_id": "',
+                AuroraSdk.currentAccountId(),
+                '",',
+                '"amount": "',
+                Strings.toString(refundAmount),
+                '",',
+                '"msg": "',
+                _toHexString(uint160(sender), 20),
+                '"}'
+            );
+            PromiseCreateArgs memory callFtTransfer = near.call(
+                tokenIdOnNear,
+                "ft_transfer_call",
+                data,
+                1,
+                REFUND_NEAR_GAS
+            );
+            callFtTransfer.transact();
```

Comment:
What if this function will fail? For example because we don't have 1 yoctowNEAR.

Reply:
If this function fails then the funds would remain in the contract's XCC account. It would be possible to add an owner-only manual recovery as a fallback option.